### PR TITLE
Div typescript improvements

### PR DESCRIPTION
**Pull request by eikhr**
# Description

Improved typings in some form stuff, some router stuff and rewrote the `LoginPage` component to a functional component

# Result

Hopefully no functional changes

### <samp>🤖 Generated by Copilot at d36e516</samp>

> _The pull request adds types galore_
> _To make the code safer and more_
> _It also refactors_
> _Some components and factors_
> _The `isCurrentUser` check in the core_

# Testing

- [x] I have thoroughly tested my changes.

I have tested the functionality of the components I made changes to, and in my testing it still works exactly the same:)

### CODE REVIEW

**Review by LudvigHz — APPROVED** [review]
LGTM

File: app/components/Form/LegoFinalForm.tsx [hunk 1]
```diff
@@ -42,8 +35,14 @@ const LegoFinalForm = <FormValues,>({
       {...rest}
       initialValuesEqual={isEqual}
       decorators={decorators}
-      onSubmit={(values, form) =>
-        onSubmit(values, form).catch((error) => {
+      onSubmit={(values, form) => {
+        const res = onSubmit(values, form);
+
+        if (!res || !('then' in res)) {
```
Comment (ivarnakken) [inline app/components/Form/LegoFinalForm.tsx:41]: How does `'then' in res` work? Is there really no other way?

**Review by ivarnakken — APPROVED** [review]
Good shit 💯

**Review by norbye — APPROVED** [review]
LGTM

File: app/routes/users/utils.ts [hunk 2]
```diff
@@ -17,3 +19,11 @@ export const validPassword =
...
     return [evalPass.score >= 3, message];
   };
+
+export const isCurrentUser = (username: string, currentUsername: string) =>
+  username === 'me' || username === currentUsername;
+
+export const useIsCurrentUser = (username: string) => {
+  const { currentUser } = useUserContext();
+  return isCurrentUser(username, currentUser.username);
+};
```
Comment (ivarnakken) [inline app/routes/users/utils.ts:29]: Nice this is very useful!

File: app/routes/users/components/UserProfile.tsx [hunk 3]
```diff
@@ -395,9 +395,9 @@ const UserProfile = (props: Props) => {
             {hasFrame && (
               <Image alt="Golden frame" className={styles.frame} src={frame} />
             )}
-            <ProfilePicture user={user} size={150} />
+            <ProfilePicture user={user} alt="Profile picture" size={150} />
```
Comment (ivarnakken) [inline app/routes/users/components/UserProfile.tsx:398]: A `<ProfilePicture />` component shouldn't really need a `alt` prop as it should always be `{user.fullname}'s profile picture` or something.

File: app/routes/users/utils.ts (hunk 2 above)
Comment (ivarnakken) [inline app/routes/users/utils.ts:29]: Very cool and much needed!

**Review by ivarnakken — APPROVED** [review]

**Pull request merged**